book/src/designs/0002-parallel-verification.md

```diff
@@ -0,0 +1,426 @@
+    storing any intermediate values in the context.
+* **Heartwood:**
+  * Heartwood changes the meaning of the `history_root_hash` field in the block
+    header. For Sapling and Blossom, it is the final sapling treestate hash.
+    For Heartwood, it is the root hash of a Merkle Mountain Range (MMR) which
+    commits to various features of the chain history. See
+    [ZIP-221](https://zips.z.cash/zip-0221) for details.
+  * Each new block adds a leaf node, then merges subtree roots, if possible.
+    It generates some new "extra nodes" to bag the subtrees, then calculates
+    the root hash.
+  * To efficiently verify this field, Zebra can store the list of MMR previous
+    subtree roots in the `ChainContext` as an `im::Vector`. The number of
+    subtree roots is at most `log(height - activation_height)`.
+  * As an optimisation, we could also store a list of recent "extra nodes"
+    generated during the bagging process.
+  * Store the full list of previous subtree roots in the `DiskState`. Update by
+    replacing the old `DiskState` list with the new `ChainContext` set.
+  * The Heartwood activation block has an all-zeroes `history_root_hash` field,
+    and an empty list of previous subtree roots.
+
+## Main Chain Tip
+[main-chain-tip]: #main-chain-tip
+
+The main chain tip determines the set of valid, historical Zcash transactions.
+The main chain tip is updated whenever a new block is received:
+
+First, the set of chain tips is updated:
+* if the new block uniquely extends an existing chain tip, that existing chain
+  tip is replaced by the new block.
+* if the new block forks from the ancestor of an existing tip, the new block
+  is added to the set of chain tips.
+
+Then, the new main tip is selected, according to these rules:
+* the main tip is the chain tip with the greatest cumulative proof of work,
+  calculated according to the Zcash Specification. (This is a consensus rule.)
+* as a tie-breaker, if multiple chain tips have equal cumulative work, and one
+  of those tips is the current main tip, the main tip does not change. (This is
+  *not* a consensus rule, because it depends on download and verification
+  order on each local node.)
+  * Note: zcashd chooses the first block that was downloaded on the local
+    node. But in Zebra, we want to avoid tracking an associated download time
+    for each block (in memory and on disk).
+* as a tie-breaker, if the main tip is not one of the chain tips with the
+  greatest cumulative work, the chain tip with the lowest BlockHeaderHash
+  becomes the main chain tip. (This is *not* a consensus rule. But just in
+  case we want to turn it into a consensus rule in future, we specify that
+  the comparison should happen in little-endian byte order.)
+  * Note: Since the `ChainTipUpdater` has exclusive access to the chain tips,
+    this should be impossible, unless a network upgrade changes the proof of
+    work rules, or we store blocks without associating them with a chain.
```

Comment:
```suggestion
  greatest cumulative work, the service chooses an arbitrary chain tip.
  (This is *not* a consensus rule, and it should not affect network
  convergence, because that is handled by the previous two rules.)
   * Note: we can avoid this edge case by making sure that all verified
     blocks are associated with a chain. This ensures they are an
     ancestor of at least one tip. (Or the tip itself.)
  * Note: Since the service has exclusive access to the chain tips, and
    it only adds one block at a time, this edge case should be
    impossible.
  * Note: if a network upgrade changes the proof of work rules, it
     could cause a tie. We should review this design if the proof of
     work rules change.
```